source/interface/azure_iot_mqtt.h

```diff
@@ -0,0 +1,380 @@
+} AzureIoTMQTTConnectInfo_t;
+
+typedef struct AzureIoTMQTTSubscribeInfo
+{
+    /**
+     * @brief Quality of Service for subscription.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Topic filter to subscribe to.
+     */
+    const char * pcTopicFilter;
+
+    /**
+     * @brief Length of subscription topic filter.
+     */
+    uint16_t usTopicFilterLength;
+} AzureIoTMQTTSubscribeInfo_t;
+
+typedef struct AzureIoTMQTTPublishInfo
+{
+    /**
+     * @brief Quality of Service for message.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Whether this is a retained message.
+     */
+    bool xRetain;
+
+    /**
+     * @brief Whether this is a duplicate publish message.
+     */
+    bool xDup;
+
+    /**
+     * @brief Topic name on which the message is published.
+     */
+    const char * pcTopicName;
+
+    /**
+     * @brief Length of topic name.
+     */
+    uint16_t usTopicNameLength;
+
+    /**
+     * @brief Message payload.
+     */
+    const void * pvPayload;
+
+    /**
+     * @brief Message payload length.
+     */
+    size_t xPayloadLength;
+} AzureIoTMQTTPublishInfo_t;
+
+typedef struct AzureIoTMQTTDeserializedInfo
+{
+    uint16_t usPacketIdentifier;                 /** @brief Packet ID of deserialized packet. */
+    AzureIoTMQTTPublishInfo_t * pxPublishInfo;   /** @brief Pointer to deserialized publish info. */
+    AzureIoTMQTTResult_t xDeserializationResult; /** @brief Return code of deserialization. */
+} AzureIoTMQTTDeserializedInfo_t;
+
+typedef struct AzureIoTMQTTPacketInfo
+{
+    /**
+     * @brief Type of incoming MQTT packet.
+     */
+    uint8_t ucType;
+
+    /**
+     * @brief Remaining serialized data in the MQTT packet.
+     */
+    uint8_t * pucRemainingData;
+
+    /**
+     * @brief Length of remaining serialized data.
+     */
+    size_t xRemainingLength;
+} AzureIoTMQTTPacketInfo_t;
+
+typedef AzureIoTMQTT_t * AzureIoTMQTTHandle_t;
+
+typedef uint32_t ( * AzureIoTMQTTGetCurrentTimeFunc_t )( void );
+
+typedef void ( * AzureIoTMQTTEventCallback_t )( AzureIoTMQTTHandle_t pContext,
+                                                struct AzureIoTMQTTPacketInfo * pxPacketInfo,
+                                                struct AzureIoTMQTTDeserializedInfo * pxDeserializedInfo );
+
+
+/**
+ * @brief Initialize an AzureIoTMQTT context.
+ *
+ * @param[in] xContext The context to initialize.
+ * @param[in] pxTransportInterface The transport interface to use with the context.
+ * @param[in] xGetTimeFunction The time utility function to use with the context.
+ * @param[in] xUserCallback The user callback to use with the context to
+ * notify about incoming packet events.
+ * @param[in] pucNetworkBuffer Network buffer provided for the context.
+ * @param[in] ulNetworkBufferLength Length of network buffer.
+ *
+ * @return An #AzureIoTMQTTResult_t with the result of the operation.
+ *      - eAzureIoTMQTTBadParameter if invalid parameters are passed.
```

Comment:
Are we sure we want to explicitly document these error codes?  What happens if we discover a new code that we could return over time - these are the type of things that we tend to forget to update docs for.  When it says "eMQTTSuccess otherwise" when the BadParemeter is onyl other doc'd option I think is over-promising.

Reply:
I can remove this additional error and only state :
If operation succeeds return eAzureIoTMQTTSuccess

---
In reply to: [598962652](https://github.com/Azure/azure-iot-middleware-freertos/pull/44#discussion_r598962652) [](ancestors = 598962652)